service-netty/src/main/java/info/bitrich/xchangestream/service/netty/NettyStreamingService.java

```diff
@@ -312,28 +328,27 @@ public void sendMessage(String message) {
             if (webSocketChannel == null || !webSocketChannel.isOpen()) {
                 e.onError(new NotConnectedException());
             }
-
-            if (!channels.containsKey(channelId)) {
+            channels.computeIfAbsent(channelId, chanId -> {
```

Comment:
Looks like `sendMessage` is non-blocking, but anyway, I think, this computation can be safely moved outside of lambda. So, it can be rewritten like this:

```
Subscription newSubscription = new Subscription(e, channelName, args);
if (channels.putIfAbsent(channelId, newSubscription) == null) {
    try {
        sendMessage(getSubscribeMessage(channelName, args));
    } catch (IOException throwable) {
        e.onError(throwable);
    }
}
```

The reason for this is that underlying lock will be taken for shorter period of time.